### PR TITLE
Mock pkcs11 lib and add tests for existing API

### PR DESCRIPTION
**Pull request by danixeee**
Implemented dummy pkcs11 library in order to test existing API functions.
Tests running on _travis_ do not load real `PyKCS11Lib`, instead they are using mocked one.

### CODE REVIEW

File: taf_sc/api.py [hunk 1]
```diff
@@ -1,37 +1,39 @@
...
+  return bool(pkcs11.getSlotList(tokenPresent=True))
 
 
 @contextmanager
-def sc_session(pin):
+def sc_session(pin, pkcs11=PKCS11):
   """Try to login with provided PIN and return session."""
-  if not sc_is_present():
-    raise SmartCardNotPresent('Please insert your smart card.')
+  if not sc_is_present(pkcs11=pkcs11):
```
Comment (renatav) [inline taf_sc/api.py]: `sc_is_presen(pkcs11) is enough since `sc_is_present` does not have any other parameters`

File: taf_sc/api.py [hunk 2]
```diff
@@ -1,37 +1,39 @@
...
+def sc_is_present(pkcs11=PKCS11):
   """Check if smart card is inserted."""
-  return bool(PKCS11.getSlotList(tokenPresent=True))
+  return bool(pkcs11.getSlotList(tokenPresent=True))
 
 
 @contextmanager
-def sc_session(pin):
+def sc_session(pin, pkcs11=PKCS11):
   """Try to login with provided PIN and return session."""
```
Comment (renatav) [inline taf_sc/api.py]: Typo login -> log in

File: taf_sc/exceptions.py [hunk 3]
```diff
@@ -6,9 +6,20 @@ class SmartCardError(Exception):
   pass
 
 
-class SmartCardNotPresent(SmartCardError):
+class SmartCardFindObjectError(SmartCardError):
```
Comment (renatav) [inline taf_sc/exceptions.py]: I would mention the "key" in the name of this class, like `SmardCardFindKeyObjectError`. FindObject sounds very generic

File: tests/pkcs11.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def _is_valid_mechanism(mechanism):
+  return mechanism._mech.mechanism == VALID_MECH._mech.mechanism and \
+      mechanism._param.hashAlg == VALID_MECH._param.hashAlg and \
+      mechanism._param.mgf == VALID_MECH._param.mgf and \
+      mechanism._param.sLen == VALID_MECH._param.sLen
+
+
+class _Session:
```
Comment (renatav) [inline tests/pkcs11.py:15]: Could you add some docstrings?

File: tests/test_api.py [hunk 5]
```diff
@@ -1,3 +1,69 @@
...
+
+@pytest.mark.skip_smartcard
+@pytest.mark.parametrize('pkcs11', [dict(sc_inserted=False)], indirect=True)
+def test_sc_is_present_should_return_false(pkcs11):
+  is_present = sc_is_present(pkcs11=pkcs11)
+  assert not is_present
+
+
+def test_sc_session_with_valid_pin_should_return_session_obj(pkcs11):
+  with sc_session(VALID_PIN, pkcs11=pkcs11) as session:
```
Comment (renatav) [inline tests/test_api.py:24]: `sc_session(VALID_PIN, pkcs11)` should work

**Review by renatav — CHANGES_REQUESTED** [review]
Looks good. A couple of comments.

File: tests/pkcs11.py (hunk 4 above)
Comment (danixeee) [reply tests/pkcs11.py:15]: I have added a link to `PyKCS11` session implementation.

**Comment by danixeee** [comment]
@renatav For some reason, `PyKCS11Lib` needs to be re-instantiated before each API call when using this library in a long-running process (e.g. server). Please review again.

**Comment by danixeee** [comment]
@renatav I have changed project name to `oll-sc`. Please delete it locally and clone it again.

File: oll_sc/__init__.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  """Decorator to reinitialize PyKCS11 lib. Needed for long running processes.
+  """
+  @wraps(api_func)
+  def decorator(*args, **kwargs):
+    """If pkcs11 is NOT passed in kwargs, instantiate it and add it to
+    kwargs.
+                  NOTE: pkcs11 MUST be passed as kwarg!
+    """
+    pkcs11 = kwargs.pop('pkcs11', None)
+    if not pkcs11:
```
Comment (renatav) [inline oll_sc/__init__.py]: It's better to use is None

**Review by renatav — COMMENTED** [review]
Looks good. One comment

**Review by renatav — APPROVED** [review]

**Pull request merged**